File: HerançaMultipla.go
package main

import "fmt"

type animal struct {
	nome string
}

func (a animal) EmitirSom() {

}

type Mamifero struct {
	animal
}

func (m Mamifero) Amamentar() {
	fmt.Printf("%s está amamentando", m.nome)
}

type Ave struct {
	animal
}

func (a Ave) Voar() {
	fmt.Printf("%s está voando", a.nome)
}

type Morcego struct {
	Mamifero
	Ave
}

func (m Morcego) EmitirSom() {
	fmt.Println("Morcego fazendo ruídos noturnos.")
}

func main() {
	morcego := Morcego{
		Mamifero: Mamifero{animal{"Batemane"}},
		Ave:      Ave{animal{"Batemane"}},
	}

	morcego.EmitirSom()
	morcego.Amamentar()
	morcego.Voar()
}

File: HerançaMultipla.java
// Classe base Animal
class Animal {
    protected String nome;

    public Animal(String nome) {
        this.nome = nome;
    }

    public void emitirSom() {
    }
}

interface Ave {
    void voar();
}

class Mamifero extends Animal {

    public Mamifero(String nome) {
        super(nome);
    }

    public void amamentar() {
        System.out.println(this.nome + "está amamentando");
    }
}
class Morcego extends Mamifero implements Ave {

    public Morcego(String nome) {
        super(nome);
    }

    public void emitirSom() {
        System.out.println("Morcego fazendo ruídos noturnos.");
    }

    public void voar() {
        System.out.println(this.nome + "está voando");
    }
}

public class HerançaMultipla {
    public static void main(String[] args) {
        Morcego morcego = new Morcego("Batman");
        morcego.emitirSom();
        morcego.amamentar();
        morcego.voar();
    }
}
